tree.go

```diff
@@ -957,8 +1073,77 @@ func (n *LeafNode) Commitment() *Point {
 	return n.commitment
 }
 
-func (n *LeafNode) Commit() *Point {
-	return n.commitment
+var frPool = sync.Pool{
+	New: func() any {
+		ret := make([]Fr, NodeWidth)
+		return &ret
+	},
+}
+
+func (leaf *LeafNode) Commit() *Point {
+	// If we've never calculated a commitment for this leaf node, we calculate the commitment
+	// in a single shot considering all the values.
+	if leaf.commitment == nil {
```

Comment:
could that not be simplified by simply setting the commitment to zero and then applying the CoW logic below to it? This should be equivalent, unless I missed something?

Reply:
Yes, I went with that approach initially since it would avoid an "if" case here. Unfortunately, that was slower.

The reason for that is that if you initialize a leaf with 10 values, that would mean doing 10 diff-updating instead of a single polynomial commitment with 10 values.

If you have a calculated commitment and 10 values to update with diffs-commits, you have to do 10 diff commitment which is slower than doing a single 10-values polynomial commitment.